File: Day4/PerformanceComp3/cpp_project/sort_benchmark.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>

int main() {
    const int NUM_ELEMENTS = 50'000'000;
    std::cout << "Generating " << NUM_ELEMENTS << " random integers..." << std::endl;

    std::vector<int> data(NUM_ELEMENTS);
    
    // Fast PRNG for generating data
    std::mt19937 rng(42); // Fixed seed for reproducible data
    std::uniform_int_distribution<int> dist;
    
    for (int i = 0; i < NUM_ELEMENTS; ++i) {
        data[i] = dist(rng);
    }

    std::cout << "Starting C++ std::sort..." << std::endl;

    auto start_time = std::chrono::high_resolution_clock::now();
    
    std::sort(data.begin(), data.end());
    
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end_time - start_time;

    std::cout << "Sorting complete!" << std::endl;
    std::cout << "C++ Sorting Time: " << diff.count() << " seconds" << std::endl;
    
    // Just print something to prevent over-optimization
    std::cout << "First 5 elements: ";
    for(int i=0; i<5; ++i) std::cout << data[i] << " ";
    std::cout << "\nLast 5 elements: ";
    for(int i=NUM_ELEMENTS-5; i<NUM_ELEMENTS; ++i) std::cout << data[i] << " ";
    std::cout << std::endl;

    return 0;
}

File: Day4/PerformanceComp3/rust_project/src/bin/sort_benchmark.rs
use std::time::Instant;

// Simple fast PRNG so we don't need external crates
struct XorShift {
    state: u32,
}
impl XorShift {
    fn new(seed: u32) -> Self { Self { state: seed } }
    fn next(&mut self) -> i32 {
        let mut x = self.state;
        x ^= x << 13;
        x ^= x >> 17;
        x ^= x << 5;
        self.state = x;
        x as i32
    }
}

fn main() {
    let num_elements = 50_000_000;
    println!("Generating {} random integers...", num_elements);

    let mut data = Vec::with_capacity(num_elements);
    let mut rng = XorShift::new(42);
    
    for _ in 0..num_elements {
        data.push(rng.next());
    }

    println!("Starting Rust slice::sort_unstable...");

    let start_time = Instant::now();
    
    data.sort_unstable();
    
    let duration = start_time.elapsed();

    println!("Sorting complete!");
    println!("Rust Sorting Time: {:.4} seconds", duration.as_secs_f64());
    
    // Print something to prevent over-optimization
    print!("First 5 elements: ");
    for i in 0..5 { print!("{} ", data[i]); }
    print!("\nLast 5 elements: ");
    for i in num_elements-5..num_elements { print!("{} ", data[i]); }
    println!();
}
